content/blog/2019-10-19-loop-invariant-code-motion/index.md

```diff
@@ -0,0 +1,297 @@
+This means we have to rename jumps that target the loop header to target the preheader instead.
+We only rename jumps that target the loop header if they are outside the loop. This ensures that the preheader is only to be executed when we first enter the loop.
+
+
+### Instruction list generation
+
+
+Next is to generate a new list of instructions for the function body. To do this, we initially order the basic blocks. The order must follow several constraints: the first and last basic blocks of the original program remain in place. Fall-throughs between basic blocks and the original function are also respected.
+
+Here, we ran into a few bugs. 🐝🐛🦋🐌🐞🐜🦗🕷🦂
+
+The existing code in the repository that computes dominator trees throws an exception when a basic block is unreachable. To solve this, we removed the unreachable block from all data structures. This implements a very limited form of dead code eliminations as a bonus! ☠️
+
+When we have an order for basic blocks we concatenate them to create a new list of instructions. Further post-processing to remove empty preheaders is then executed. Finally we have our new instruction list!
+
+# Evaluation
+We evaluated our optimization on a suite of benchmarks by instrumenting the `bril` interpreter to count the number of instructions it executes. Our optimization are evaluated this way, as opposed to comparing execution times. This avoids measurement biases that might arise through timing. 
+
+By abstracting away performance in terms of the number of instructions executed, we give a fair comparison between non-optimized and optimized `bril` programs: no matter the environment in which they are executed, a `bril` program that executes fewer instructions performs better than an equivalent `bril` program that executes more instructions.
+
+
+
+
+
+|test|original|optimized|
+|:---|----:|:---|
+|`examples/df_test/cond.bril`           |9|9|
+|`examples/df_test/fact.bril`           |62|55|
+|`examples/dom_test/loopcond.bril`      |117|108|
+|`examples/lvn_test/clobber-fold.bril`  |10|10|
+|`examples/lvn_test/clobber.bril`       |10|10|
+|`examples/lvn_test/commute.bril`       |6|6|
+|`examples/lvn_test/idchain.bril`       |5|5|
+|`examples/lvn_test/idchain-prop.bril`  |5|5|
+|`examples/lvn_test/nonlocal.bril`      |7|7|
+|`examples/lvn_test/reassign.bril`      |3|3|
+|`examples/lvn_test/redundant.bril`     |6|6|
+|`examples/lvn_test/redundant-dce.bril` |6|6|
+|`examples/tdce_test/combo.bril`        |6|6|
+|`examples/tdce_test/diamond.bril`      |6|6|
+|`examples/tdce_test/double-pass.bril`  |6|6|
+|`examples/tdce_test/reassign.bril`     |6|6|
+|`examples/tdce_test/reassign-dkp.bril` |6|6|
+|`examples/tdce_test/simple.bril`       |5|5|
+|`examples/tdce_test/skipped.bril`      |4|4|
+|`examples/loop_test/depend.bril`       |158|159|
+|`examples/loop_test/fibonacci.bril`    |78|79|
+|`examples/loop_test/nest.bril`         |147|130|
+|`examples/loop_test/hoist_expr.bril`   |48|32|
```

Comment:
This is very nice and very thorough to try everything, but it would have been defensible to filter out programs without any loops (i.e., no backedges).